plugins/plugin.go

```diff
@@ -163,22 +164,23 @@ func (p *Plugin) getNodeMetrics() (producers.MetricsMessage, error) {
 // endpoint for each container on the machine. It returns a slice of metrics
 // messages, one for each hit.
 func (p *Plugin) getContainerMetrics() ([]producers.MetricsMessage, error) {
-	var messages []producers.MetricsMessage
 	ids, err := p.getContainerList()
 	if err != nil {
 		return nil, errors.Wrap(err, "could not read list of containers")
 	}
+
+	var futures []<-chan producers.MetricsMessage
 	for _, id := range ids {
-		containerMetrics, err := makeMetricsRequest(p.Client, fmt.Sprintf("http://localhost/v0/containers/%s", id))
-		if err != nil {
-			return nil, errors.Wrapf(err, "could not retrieve metrics for container %s", id)
-		}
-		containerAppMetrics, err := makeMetricsRequest(p.Client, fmt.Sprintf("http://localhost/v0/containers/%s/app", id))
-		if err != nil {
-			return nil, errors.Wrapf(err, "could not retrieve app metrics for container %s", id)
-		}
-		messages = append(messages, containerMetrics, containerAppMetrics)
+		url1 := fmt.Sprintf("http://localhost/v0/containers/%s", id)
+		url2 := fmt.Sprintf("http://localhost/v0/containers/%s/app", id)
+		futures = append(futures, makeMetricsRequestFuture(p.Client, url1), makeMetricsRequestFuture(p.Client, url2))
```

Comment:
A problem here is that you are creating an unbounded number of goroutines.  This can have resource exhaustion consequences at the client and possibly also the server due to resource exhaustion.  I'd pick a magic number and limit the concurrency to that number of goroutines.